src/Ocelot/DownstreamUrlCreator/Middleware/DownstreamUrlCreatorMiddleware.cs

```diff
@@ -98,6 +98,15 @@ private static string MergeQueryStringsWithoutDuplicateValues(string queryString
             newQueryString = newQueryString.Replace(QuestionMark, Ampersand);
             var queries = HttpUtility.ParseQueryString(queryString);
             var newQueries = HttpUtility.ParseQueryString(newQueryString);
+
+            // Remove old replaced query parameters
+            foreach (var queryKey in queries.AllKeys)
+            {
+                foreach (var unused in placeholders.Where(placeholder => queryKey == placeholder.Name.Trim(OpeningBrace, ClosingBrace)))
+                {
+                    queries.Remove(queryKey);
+                }
+            }

```

Comment:
```suggestion
           var placeholderNames = new HashSet<string>(placeholders.Select(p => p.Name.Trim(OpeningBrace, ClosingBrace)));
foreach (var queryKey in queries.AllKeys.Where(placeholderNames.Contains).ToList())
{
    queries.Remove(queryKey);
}
```

Reply:
We might win some cpu cycles here, up to you

Reply:
Why the .ToList() ?

Reply:
@bbenameur yes you can remove that, I have tried several strategies and at some point Resharper complained about multiple enumerations.

Reply:
Fixed in commit https://github.com/ThreeMammals/Ocelot/pull/2003/commits/c43030ca4dcee8be1d3be665be5bb4b535b2db75 ✔️